File: math/valid-perfect-square.py
# https://leetcode.com/problems/valid-perfect-square/
# tags - math, binary-search
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        # Apply binary search to find number that equals the given number when squared.
        low, high = 1, 2**16
        while low <= high:
            mid = (low + high)//2
            squared = mid*mid
            if squared < num:
                low = mid+1
            elif squared > num:
                high = mid-1
            else:
                return True
        return False

File: math/valid-perfect-square.js
// https://leetcode.com/problems/valid-perfect-square
// tags - math
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function (num) {
  let left = 1,
    right = num;
  while (left <= right) {
    const mid = left + Math.floor((right - left) / 2);
    const squared = mid * mid;
    if (squared === num) {
      return true;
    } else if (squared > num) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }
  return false;
};
